essentials/src/api/executor.rs

```diff
@@ -0,0 +1,495 @@
+
+enum ExecutorMessage {
+	Close,
+	Rpc(OneshotSender<Response>, Request),
+}
+
+#[derive(Debug, Clone)]
+pub enum Request {
+	GetBlockTimestamp(H256),
+	GetHead(Option<H256>),
+	GetBlockNumber(Option<H256>),
+	GetBlockHash(Option<BlockNumber>),
+	GetEvents(H256),
+	ExtractParaInherent(Option<H256>),
+	GetClaimQueue(H256),
+	GetOccupiedCores(H256),
+	GetBackingGroups(H256),
+	GetSessionIndex(H256),
+	GetSessionAccountKeys(u32),
+	GetSessionNextKeys(AccountId32),
+	GetInboundHRMPChannels(H256, u32),
+	GetOutboundHRMPChannels(H256, u32),
+	GetHostConfiguration,
+	GetBestBlockSubscription,
+	GetFinalizedBlockSubscription,
+}
+
+/// Response types for APIs.
+#[derive(Debug)]
+enum Response {
+	/// A timestamp.
+	Timestamp(Timestamp),
+	/// A block header.
+	MaybeHead(Option<Header>),
+	/// A full block.
+	BlockNumber(BlockNumber),
+	/// A block hash.
+	MaybeBlockHash(Option<H256>),
+	/// Block events
+	MaybeEvents(Option<subxt::events::Events<PolkadotConfig>>),
+	/// `ParaInherent` data.
+	ParaInherentData(InherentData),
+	/// Claim queue for parachains.
+	ClaimQueue(ClaimQueue),
+	/// List of the occupied availability cores.
+	OccupiedCores(Vec<CoreOccupied>),
+	/// Backing validator groups.
+	BackingGroups(Vec<Vec<polkadot_primitives::ValidatorIndex>>),
+	/// Returns a session index
+	SessionIndex(u32),
+	/// Session keys
+	SessionAccountKeys(Option<Vec<AccountId32>>),
+	/// Session next keys for a validator
+	SessionNextKeys(Option<SessionKeys>),
+	/// HRMP channels for some parachain (e.g. who are sending messages to us)
+	HRMPChannels(BTreeMap<u32, SubxtHrmpChannel>),
+	/// The current host configuration
+	HostConfiguration(DynamicHostConfiguration),
+	/// Chain subscription
+	ChainSubscription(HeaderStream),
+}
+
+#[derive(Debug, Error)]
+pub enum RequestExecutorError {
+	#[error("Client for url {0} not found")]
+	ClientNotFound(String),
+	#[error("subxt error: {0}")]
+	SubxtError(#[from] subxt::error::Error),
+	#[error("dynamic error: {0}")]
+	DynamicError(#[from] DynamicError),
+	#[error("subxt connection timeout")]
+	Timeout,
+	#[error("Send failed: {0}")]
+	PrioritySendError(#[from] PrioritySendError),
+	#[error("Recv failed: {0}")]
+	OneshotRecvError(#[from] OneshotRecvError),
+	#[error("Unexpected response for request {0:?}")]
+	UnexpectedResponse(Request),
+}
+
+impl RequestExecutorError {
+	pub fn should_repeat(&self) -> bool {
+		matches!(
+			self,
+			RequestExecutorError::SubxtError(subxt::Error::Io(_)) |
+				RequestExecutorError::SubxtError(subxt::Error::Rpc(_))
+		)
+	}
+}
+
+struct RequestExecutorBackend {
+	retry: RetryOptions,
+}
+
+impl RequestExecutorBackend {
+	async fn run(
+		&mut self,
+		from_frontend: PriorityReceiver<ExecutorMessage>,
+		url: String,
+		api_client_mode: ApiClientMode,
+	) {
+		let client = match build_client(&url, api_client_mode, &self.retry).await {
+			Some(v) => v,
+			None => return,
+		};
+
+		loop {
+			tokio::select! {
+				message = from_frontend.recv() => {
```

Comment:
you don't really need a `select!` here, you can just `let message = from_frontend.recv().await`